src/Builder.php

```diff
@@ -0,0 +1,246 @@
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ * This software consists of voluntary contributions made by many individuals
+ * and is licensed under the MIT license. For more information, see
+ * <http://www.doctrine-project.org>.
+ */
+
+declare(strict_types=1);
+
+namespace Doctrine\Annotations;
+
+use Doctrine\Annotations\Annotation\Target;
+use Doctrine\Annotations\Metadata\ClassMetadata;
+use Doctrine\Annotations\Metadata\MetadataFactory;
+use Doctrine\Annotations\Exception\TypeMismatchException;
+use Doctrine\Annotations\Exception\TargetNotAllowedException;
+use Doctrine\Annotations\Exception\InvalidAnnotationException;
+
+/**
+ * Build annotations objects
+ *
+ * @author Fabio B. Silva <fabio.bat.silva@gmail.com>
+ */
+class Builder
+{
+    /**
+     * @var \Doctrine\Annotations\Resolver
+     */
+    private $resolver;
+
+    /**
+     * @var \Doctrine\Annotations\MetadataFactory
+     */
+    private $metadataFactory;
+
+    /**
+     * Constructor.
+     *
+     * @param \Doctrine\Annotations\Resolver        $resolver
+     * @param \Doctrine\Annotations\MetadataFactory $metadataFactory
+     */
+    public function __construct(Resolver $resolver, MetadataFactory $metadataFactory)
+    {
+        $this->resolver        = $resolver;
+        $this->metadataFactory = $metadataFactory;
+    }
+
+    /**
+     * @return \Doctrine\Annotations\Resolver
+     */
+    public function getResolver() : Resolver
+    {
+        return $this->resolver;
+    }
+
+    /**
+     * @param Context   $context
+     * @param Reference $reference
+     *
+     * @return object
+     */
+    public function create(Context $context, Reference $reference)
+    {
+        $target    = $reference->nested ? Target::TARGET_ANNOTATION : $context->getTarget();
+        $fullClass = $this->resolver->resolve($context, $reference->name);
+        $metadata  = $this->metadataFactory->getMetadataFor($fullClass);
+        $values    = $reference->values;
+
+        if ($metadata === null) {
+            throw InvalidAnnotationException::notAnnotationException($fullClass, $reference->name, $context->getDescription());
+        }
+
+        if (($metadata->target & $target) === 0) {
+            $contextDesc   = $context->getDescription();
+            $allowedTarget = implode(',', Target::getNames($metadata->target));
+
+            throw TargetNotAllowedException::notAllowedDeclaration($fullClass, $contextDesc, $allowedTarget);
+        }
+
+        return $this->instantiate($context, $metadata, $values);
+    }
+
+    /**
+     * @param Context       $context
+     * @param ClassMetadata $metadata
+     * @param array         $values
+     *
+     * @return object
+     */
+    private function instantiate(Context $context, ClassMetadata $metadata, array $values)
```

Comment:
I'd argue the instantiation of an annotation should be moved to another class, `Instantiator`.
However, this would defeat the purpose of having a `Builder` class. Would love to hear pros/cons of eliminating this class and merging the `create()` method into `DocVisitor` .